test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -60,6 +85,27 @@
         ;; make sure we got here
         (is (true? true))))))
 
+(deftest pool-lock-is-blocking-test-until-borrows-unregistered-test
+  (let [pool (create-populated-pool 3)
+        instances (borrow-n-instances pool 2)
+        future-started? (promise)]
+    (is (= 2 (count instances)))
+    (let [lock-thread (future (deliver future-started? true)
+                              (.lock pool))]
+      @future-started?
+      (testing "pool.lock() blocks until all instances are unregistered"
+        (is (not (realized? lock-thread)))
+
+        (testing (str "other threads may successfully unregister instances "
+                      "while pool.lock() is being executed")
+          (.unregister pool (first instances))
+          (is (not (realized? lock-thread)))
+          (.unregister pool (second instances)))
+
+        @lock-thread
+        ;; make sure we got here
+        (is (true? true))))))
```

Comment:
> Yeah, I'm definitely good with deferring any additional changes, just trying to make sure I understand the current semantics. If you're confident that, after the final unregister, the lock will be granted

I added the suggested test from earlier in c5d08f to confirm that this is the case.

> and there will be a poison pill...

I didn't add anything to this PR to make that happen, but this line of code already in the master branch appears to do that:

https://github.com/puppetlabs/puppet-server/blob/puppet-server-2.2.1/src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj#L122

--

After some additional IRL discussion about this, I filed a follow-up ticket that would cover looking into this further - [PE-13552](https://tickets.puppetlabs.com/browse/PE-13552).  That one will probably need to have a SERVER project ticket spun off of it as well for the open-source work but thought it would be better to start with it as a PE ticket to research the interaction between the pool flush and file-sync pool locking.